flang/lib/Lower/IO.cpp

```diff
@@ -196,6 +196,144 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
   builder.setInsertionPointToStart(&ifOp.thenRegion().front());
 }
 
+/// Retrieve or generate (the constant parts of) information for NAMELIST
+/// group `symbol`.  Group object descriptor addresses are inserted at
+/// runtime.  The data structure containing the information is defined in
+/// runtime header file namelist.h as:
+///
+///     class NamelistGroup {
+///     public:
+///       struct Item {
+///         const char *name; // NUL-terminated lower-case
+///         const Descriptor &descriptor;
+///       };
+///       const char *groupName; // NUL-terminated lower-case
+///       std::size_t items;
+///       const Item *item; // in original declaration order
+///     };
+///
+/// Return the Item array and the NamelistGroup data structure.
+static std::pair<fir::GlobalOp, fir::GlobalOp>
```

Comment:
Is this really OK for these structures to be globals modified before IO calls using them ?
I guess re-entrency is not much of an issue given it I do not think the IO statement can contain a call that itself uses the same namelist, but wouldn't it be an issue when mixing with OpenMP/OpenACC ?

Does nvfortran support namelist IO in parallel contexts ?

Reply:
I've switched to regenerating all information for each use of a namelist group.  The net result is that only constant name strings are reused.